### PR TITLE
Support members for the new newtypes

### PR DESCRIPTION
**Pull request by RustanLeino**
This PR adds support for members of the recently expanded newtypes. In particular, inheritance of user-defined members and built-in members (such as `.RotateLeft` for bitvector types) are handled. Under `--general-trait=full`, inheritance from both `newtype` base types and parent traits are handled.

A `newtype`'s inherited members can only be figured out once the `newtype`'s base type is known. Since the base type is in general inferred, this means that member registration cannot happen until resolution has started. So, this PR moves that process from `Register...` to `Resolve...`.

As part of this PR, other built-in types (`char`, `set`, ...) were added to the `valuetypeDecls` array. This has an effect on the `--rprint` output.

The PR also improves some error messages (e.g., making them more consistent by using the format `type 'T'`) and removes some unused code that was left in abandoned `#if ...` directives during the resolver refresh.

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: Source/IntegrationTests/TestFiles/LitTests/LitTest/dafny0/TypeConstraints.dfy.expect [hunk 1]
```diff
@@ -110,6 +112,14 @@ module _System {
...
+  type set<+T> { }
+
+  type iset<*T> { }
+
+  type seq<+T> { }
+
+  type multiset<+T> { }
+
   type map<+T, +U> {
     var Keys: set<T>  // immutable
```
Comment (MikaelMayer) [inline Source/IntegrationTests/TestFiles/LitTests/LitTest/dafny0/TypeConstraints.dfy.expect:124]: Immutable with var? Shouldn't it be const?

File: Source/DafnyCore/Resolver/PreType/PreTypeResolve.cs [hunk 2]
```diff
@@ -405,7 +387,7 @@ public enum Type2PreTypeOption { GoodForInference, GoodForPrinting, GoodForBoth
     public static bool IsArrayName(string name, out int dimensions) {
       Contract.Requires(name != null);
       if (name.StartsWith(PreType.TypeNameArray)) {
-        var dims = name.Substring(5);
+        var dims = name[5..];
```
Comment (MikaelMayer) [inline Source/DafnyCore/Resolver/PreType/PreTypeResolve.cs:390]:
```suggestion
        var dims = name[PreType.TypeNameArray.Length..];
```

**Review by MikaelMayer — COMMENTED** [review]
A few comments here and there, but I'm happy with this very well tested PR otherwise!

File: Source/DafnyCore/Resolver/ModuleResolver.cs [hunk 3]
```diff
@@ -1021,22 +1022,6 @@ public partial class ModuleResolver : INewOrOldResolver {
...
       }
-      if (prevErrorCount == reporter.Count(ErrorLevel.Error)) {
-        // Register the trait members in the classes that inherit them
-        foreach (TopLevelDecl d in declarations) {
-          if (d is TopLevelDeclWithMembers cl) {
-            RegisterInheritedMembers(cl);
-          }
-        }
-      }
-      if (prevErrorCount == reporter.Count(ErrorLevel.Error)) {
```
[the remark this answers is not in the document]
Comment (RustanLeino) [reply Source/DafnyCore/Resolver/ModuleResolver.cs:1032]: Thanks for noticing. Yes, this is as intended. The `RegisterInheritedMembers` and `InheritedTraitMembers` methods had been responsible for bringing in the members from a type's trait parents. With the more general `newtype`s, a `newtype` can now inherit members also from its base type. However, the base type is, in general, not known this early in the resolution pipeline. Therefore, the functionality of `RegisterInheritedMembers` and `InheritedTraitMembers` was moved into "pass 0" of the resolver.

File: Source/DafnyCore/Resolver/ModuleResolver.cs [hunk 4]
```diff
@@ -2157,6 +2164,81 @@ public ReportOtherAdditionalInformation_Visitor(ModuleResolver resolver)
           clMember.OverriddenMember = traitMember;
         }
       }
+      ProcessInheritedTraitMembers(cl, membersWithErrors);
+    }
+
+    private void AddToInheritedMembers(TopLevelDeclWithMembers cl, TopLevelDeclWithMembers baseOrParentTypeDecl,
+      Dictionary<string, MemberDecl> inheritedMembers, List<string> membersWithErrors) {
+      // TODO: in the following line, rather than using .Values, it would be nice to use something that gave a deterministic order
+      foreach (var inheritedMember in GetClassMembers(baseOrParentTypeDecl)!.Values) {
```
[the remark this answers is not in the document]
Comment (RustanLeino) [reply Source/DafnyCore/Resolver/ModuleResolver.cs]: Thank you!

File: Source/IntegrationTests/TestFiles/LitTests/LitTest/dafny0/TypeConstraints.dfy.expect (hunk 1 above)
Comment (RustanLeino) [reply Source/IntegrationTests/TestFiles/LitTests/LitTest/dafny0/TypeConstraints.dfy.expect:124]: Good call. I changed the printing code to output these `SpecialField`s as `const`.

**Review by MikaelMayer — APPROVED** [review]
Looks great ! Approved a priori. Did you also ensure that the type characteristics were matched?

**Comment by RustanLeino** [comment]
> Did you also ensure that the type characteristics were matched?

No new cases needed to be checked. Unlike members inherited from a parent trait, the members that are inherited from the base type already have a body and cannot be overridden.

**Pull request merged**